src/Lumper.UI/Updater/Updater.cs

```diff
@@ -0,0 +1,362 @@
+            }
+            else
+            {
+                int read;
+                var length = (int)response.Content.Headers.ContentLength.Value;
+                var remaining = length;
+                while (!handler.Cancelled &&
+                       (read = await downloadStream.ReadAsync(
+                           buffer.AsMemory(0, int.Min(buffer.Length, remaining)),
+                           cts.Token)) >
+                       0)
+                {
+                    var prog = (float)read / length * DownloadProgressOverallProportion;
+                    handler.UpdateProgress(prog, $"{float.Floor((1 - ((float)remaining / length)) * DownloadProgressOverallProportion)}%");
+                    await stream.WriteAsync(buffer.AsMemory(0, read), cts.Token);
+                    remaining -= read;
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            if (ex is TaskCanceledException)
+                Logger.Info("Download cancelled by user");
+            else
+                Logger.Error(ex, "Failed to download!");
+
+            await stream.DisposeAsync();
+            return false;
+        }
+        finally
+        {
+            stream.Close();
+        }
+
+        if (handler.Cancelled)
+            return false;
+
+        return true;
+    }
+
+    /// <summary>
+    /// Returns the URL to the download link for the OS-specific version.
+    /// </summary>
+    /// <returns></returns>
+    private static string GetReleaseDownloadUrl(GithubUpdate assets, string osName)
+    {
+        foreach (Asset asset in assets.Assets)
+        {
+            if (asset.Name.Contains(osName, StringComparison.OrdinalIgnoreCase))
+            {
+                return asset.BrowserDownloadUrl;
+            }
+        }
+
+        throw new InvalidDataException($"Could not find download link for {osName}");
+    }
+
+    /// <summary>
+    /// Downloads an update for the program, applies it, and then restarts itself with the new version
+    /// </summary>
+    /// <returns></returns>
+    public static async ValueTask Update()
+    {
+        try
+        {
+            GithubUpdate assets = await FetchGithubUpdates();
+            LumperVersion latest = ParseVersion(assets.TagName);
+
+            OSPlatform os;
+            string osName;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                os = OSPlatform.Windows;
+                osName = "win";
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                os = OSPlatform.Linux;
+                osName = "linux";
+            }
+            else
+            {
+                throw new ApplicationException("OS is not supported");
+            }
+
+            var downloadUrl = GetReleaseDownloadUrl(assets, osName);
+            var fileName = $"{osName}_{latest}.zip";
+            var directoryName = $"{fileName}temp";
```

Comment:
write the download to `/tmp` so it will be cleaned up if something goes wrong and we don't make it to the `rm` in code. There should be a c# function to get the temp folder or a temp file independent of the platform we're on.

Reply:
Wasn't sure how long this is guaranteed to last, turns out on Windows, the OS doesn't clean up the files for you! I was under the impression they were wiped on reboot, which AFAIK is what all unix-like systems with a `/tmp` do. So, we should still `rm` the file after.